### PR TITLE
Enabled masked methods for bfloat16

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24224 Enabled torch.mm and torch.mv for bfloat16 on CPU
* **#24183 Enabled masked methods for bfloat16**
-----------
Fix: Enabled masked select/scatter/fill  for BFloat16 on CPU
Test: via unit tests

Differential Revision: [D16763461](https://our.internmc.facebook.com/intern/diff/D16763461)

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -8409,29 +8409,33 @@ def test_scatter_add_bool(self):
...
-                src = torch.randn(num_dest)
-                dest.masked_scatter_(mask, src)
-
-                # make src smaller. this should fail
-                src = torch.randn(num_copy - 1)
-                with self.assertRaises(RuntimeError):
+            for maskType in [torch.uint8, torch.bool]:
+                for dt in torch.testing.get_all_dtypes():
+                    if dt == torch.half or dt == torch.bool:
+                        continue
```
Comment (gchanan) [inline test/test_torch.py]: first, do half and double not actually work for masking?

second, as has been mentioned in a number of reviews, don't skip tests, assert they don't work with a message saying to enable once it works.  As you've done it here, we will never actually run these tests, even when we enable bool and half.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: test/test_torch.py (hunk 1 above)
Comment (izdeby) [reply test/test_torch.py]: double?

Comment (ezyang) [reply test/test_torch.py]: (probably meant bool)

File: test/test_torch.py [hunk 2]
```diff
@@ -8454,29 +8454,40 @@ def test_scatter_add_bool(self):
...
+                for dt in torch.testing.get_all_dtypes():
+                    num_copy, num_dest = 3, 10
+                    dest = torch.tensor([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=dt)
+                    dest2 = dest.clone()
+                    src = torch.tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=dt)
+                    mask = torch.tensor((0, 0, 0, 0, 1, 0, 1, 0, 1, 0), dtype=maskType)
+
+                    if dt == torch.bool:
+                        # torch.bool is a special case and is being tested
+                        # in a separate test
```
Comment (ezyang) [inline test/test_torch.py]: ...because bool will truncate elements outside of its range? (Spell it out for the reader ;)

**Review by ezyang — APPROVED** [review]

File: test/test_torch.py (hunk 2 above)
Comment (gchanan) [reply test/test_torch.py]: ...but that's not actually all that special :).  I don't know why we just don't have an if statement that sets up the destination for the case where it's bool -- we don't test the edge cases for bool tensors for what seems like non-compelling reasons.

File: test/test_torch.py [hunk 3]
```diff
@@ -8454,29 +8454,40 @@ def test_scatter_add_bool(self):
 
     def test_masked_scatter(self):
         with warnings.catch_warnings(record=True) as w:
```
Comment (gchanan) [inline test/test_torch.py]: unconditionally catching the warnings and not checking them doesn't seem great -- we don't actually check that we raise warnings for uint8 and if we start raising warnings for bool no one will notice.

File: test/test_torch.py (hunk 2 above)
Comment (izdeby) [reply test/test_torch.py]: I find this a bit more cleaner. in case when there are a lot of if/else conditions, its hard to figure out what does the test do quickly. But if you insist, i will combine all masked_* tests with their masked_*_bool versions
